kivakit-core/src/main/java/com/telenav/kivakit/core/vm/ShutdownHook.java

```diff
@@ -19,46 +19,124 @@
 package com.telenav.kivakit.core.vm;
 
 import com.telenav.kivakit.core.lexakai.DiagramLanguage;
+import com.telenav.kivakit.core.logging.Logger;
+import com.telenav.kivakit.core.logging.loggers.ConsoleLogger;
+import com.telenav.kivakit.core.string.Strings;
+import com.telenav.kivakit.core.thread.KivaKitThread;
+import com.telenav.kivakit.core.time.Duration;
 import com.telenav.lexakai.annotations.LexakaiJavadoc;
 import com.telenav.lexakai.annotations.UmlClassDiagram;
+import org.jetbrains.annotations.NotNull;
 
-import java.util.ArrayList;
-import java.util.LinkedList;
-import java.util.List;
+import java.util.PriorityQueue;
+
+import static com.telenav.kivakit.core.thread.KivaKitThread.State.EXITED;
+import static com.telenav.kivakit.core.vm.ShutdownHook.Order.MIDDLE;
+import static com.telenav.kivakit.interfaces.time.WakeState.TIMED_OUT;
 
 /**
- * Adds order-of-execution to {@link Runtime#addShutdownHook(Thread)}. Hooks can request that they be run {@link
- * Order#FIRST} or {@link Order#LAST}. The order of two hooks both requesting to be first or last is not defined. It is
- * only guaranteed that hooks asking to be run last will be run after hooks requesting to be run first.
+ * Adds <i>serial</i> execution of shutdown hooks to the functionality provided by
+ * {@link Runtime#addShutdownHook(Thread)}. When the virtual machine shuts down, the hooks registered with
+ * {@link ShutdownHook} will be called sequentially according to the ordering provided when the hooks were registered
+ * with {@link #register(String, Order, Duration, Runnable)}.
+ *
+ * <p>
+ * Hooks can request that they be run {@link Order#FIRST}, {@link Order#MIDDLE} or {@link Order#LAST}. The order of
+ * execution hooks assigned the <i>same</i> {@link Order} is not defined. Otherwise hooks will be executed in sorted
+ * order, from {@link Order#FIRST} to {@link Order#LAST}.
+ * </p>
+ *
+ * <i><b>NOTE:</b></i>
+ * <p>
+ * This class simply allows pre-registered shutdown hooks to be run in a defined order during VM shutdown. For more
+ * complex shutdown hook requirements, see <a
+ * href="https://mvnrepository.com/artifact/com.mastfrog/shutdown-hooks">com.mastfrog:shutdown-hooks</a>.
+ * </p>
  *
  * @author jonathanl (shibo)
  */
 @UmlClassDiagram(diagram = DiagramLanguage.class)
-public class ShutdownHook
+public class ShutdownHook implements Comparable<ShutdownHook>
 {
-    private static final LinkedList<ShutdownHook> queue = new LinkedList<>();
+    /** We use a console logger here because it is never involved in shutdown processes */
+    private static final Logger LOGGER = new ConsoleLogger();
+
+    /** A priority queue of shutdown hooks sorted in an assigned {@link Order} */
+    private static PriorityQueue<ShutdownHook> hooks;
```

Comment:
I still think what you want here is a pair of trieber stacks - atomics, no complex locking or ordering - faster, simpler and non-blocking for all callers.  Here is a dirt-simple trieber stack:
```
public class TrieberStack<T>
{
    private static class Cell<T>
    {
        private final T item;
        private final Cell<T> parent;

        public Cell(T item, Cell<T> parent)
        {
            this.item = item;
            this.parent = parent;
        }
        
        void fill(Collection<? super T> list) {
            list.add(item);
            if (parent != null) {
                parent.fill(list);
            }
        }
    }

    private final AtomicReference<Cell<T>> head = new AtomicReference<>();

    public void push(T obj)
    {
        head.getAndUpdate(oldHead ->
        {
            return new Cell<>(obj, oldHead);
        });
    }
    
    public T pop() {
        Cell<T> result = head.getAndUpdate(oldHead -> {
            if (oldHead == null) {
                return null;
            }
            return oldHead.parent;
        });
        return result == null ? null : result.item;
    }
    
    public void drain(Collection<? super T> coll) {
        Cell<T> cell = head.getAndSet(null);
        if (cell != null) {
            cell.fill(coll);
        }
    }
}
```

Reply:
it's a nice trick, but actually the code needs a synchronized guard around `shuttingDown` so it might as well control access to the priority queue too. and unless something went wrong (register called during shutdown, which will throw an exception), there will never be contention for this lock at all.

Reply:
Would like to go ahead and merge this. The intention of ShutdownHook is to serialize shutdown hooks so they run one after the other, in a particular order. The code contains a reference now to mastfrog for people interested in that implementation.